### PR TITLE
fix: More Stability for FormBuilder by unit tests

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

- Removed usage of 'location' from FormBuilder directly. Instead use FormBuilder props for that.
- Adds tests for FormBuilder component

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
- Verify that the location field is correctly shown as Required
- Verify that the Blue Badge is shown for Location

Follow up:
- `FormBuilderField` Tests
- More tests for FormBuilder

### CODE REVIEW

**Comment by hariombalhara** [comment]
* **#16129** <a href="https://app.graphite.dev/github/pr/calcom/cal.com/16129?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hariombalhara and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/features/form-builder/FormBuilder.test.tsx
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.test.tsx:1]: Moved all-questions tests here and renamed. Refactored a lot of stuff

File: packages/features/form-builder/FormBuilder.test.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+        await verifier.verifyFieldToggle({ identifier: newIdentifier });
+
+        await verifier.verifyFieldDeletion({ identifier: newIdentifier });
+      });
+    }
+  });
+
+  describe("radioInput type field with options available in dataStore", () => {
+    test('Should add new field of type "radioInput" and select option from dataStore', async () => {
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.test.tsx:96]: This tests location field

File: packages/features/form-builder/FormBuilder.tsx [hunk 2]
```diff
@@ -44,13 +44,6 @@ function getCurrentFieldType(fieldForm: UseFormReturn<RhfFormField>) {
   return fieldTypesConfigMap[fieldForm.watch("type") || "text"];
 }
 
-function isRequiredField(field: RhfFormField) {
-  // 'location' must be shown as required in the UI(but at backend it is not required because not specifying it defaults to CalVideo)
-  // So, handle it in UI only instead of updating bookingFields config in getBookingFields
-  // TODO: FormBuilder must not be made aware of BookingFields, so move this "location" check to outside FormBuilder where the component is used.
-  return field.name === "location" ? true : field.required;
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:51]: Acted on TODO

File: packages/features/form-builder/FormBuilder.tsx [hunk 3]
```diff
@@ -127,12 +132,13 @@ export const FormBuilder = function FormBuilder({
           {fields.map((field, index) => {
             let options = field.options ?? null;
             const sources = [...(field.sources || [])];
-            const isRequired = isRequiredField(field);
+            const isRequired = shouldConsiderRequired ? shouldConsiderRequired(field) : field.required;
             if (!options && field.getOptionsAt) {
-              options = dataStore.options[field.getOptionsAt as keyof typeof dataStore] ?? [];
-              // TODO: The dataStore should itself provide the label directly
-              // This is important because FormBuilder isn't aware of what a location is. It is a generic Form Builder
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:134]: Acted on TODO.

File: packages/features/form-builder/testUtils.ts [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+const getEditDialogForm = () => {
+  const formBuilder = document.getElementById("form-builder");
+  if (!formBuilder) {
+    throw new Error("Form Builder not found");
+  }
+  return within(formBuilder);
+};
+
+export const pageObject = {
```
Comment (hariombalhara) [inline packages/features/form-builder/testUtils.ts:71]: Holds ways to access and act on different logical elements in the component

File: packages/features/form-builder/testUtils.ts [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    return field.getByTestId("toggle-field");
+  },
+  toggleField: ({ identifier }: { identifier: string }) => {
+    const field = getFieldInTheList({ identifier });
+    fireEvent.click(field.getByTestId("toggle-field"));
+  },
+};
+
+export const verifier = {
+  verifyFieldAddition: async (props: FieldProps) => {
```
Comment (hariombalhara) [inline packages/features/form-builder/testUtils.ts:164]: It holds the same stuff that was there in now deleted utils.ts. But things have been renamed and pageObject is being used to make the code readable.

File: vitest.workspace.ts [hunk 6]
```diff
@@ -138,6 +139,15 @@ const workspaces = packagedEmbedTestsOnly
           setupFiles: ["packages/ui/components/test-setup.ts"],
         },
       },
+      {
+        test: {
+          globals: true,
+          name: "@calcom/features/form-builder",
```
Comment (hariombalhara) [inline vitest.workspace.ts:145]: form-builder needs UI environment

File: packages/prisma/schema.prisma [hunk 7]
```diff
@@ -55,6 +55,8 @@ model Host {
   @@index([eventTypeId])
 }
 
+
+
```
Comment (hariombalhara) [inline packages/prisma/schema.prisma:59]: Did this to bust DB cache by CI. Somehow something wrong was cached failing the tests.

Can remove it after the PR merge .

**Review by CarinaWolli — APPROVED** [review]
LGTM 👍

**Pull request merged**